### DISCUSSION TITLE
A question about the gzip plugin

### DISCUSSION THREAD

**Discussion by scaat** [apache/apisix#4919] [Q&A] [answered]
https://github.com/apache/apisix/blob/2.9/apisix/plugins/gzip.lua#L154
```
    local ok, err = response.set_gzip({
        buffer_num = buffers.number,
        buffer_size = buffers.size,
        compress_level = conf.comp_level,
    })
```
Where is set_gzip defined?
@spacewander thanks.

**Comment by RocFang** [comment, marked as answer]
see https://github.com/api7/apisix-nginx-module

more speicifily，https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/apisix/response.lua

**Reply by scaat** [reply]
thanks

**Reply by xlgo** [reply]
2021/12/25 05:17:37 [error] 1590#1590: *14612 failed to run header_filter_by_lua*: /usr/local/apisix/resty/apisix/response.lua:28: /usr/local/openresty/luajit/lib/libluajit-5.1.so.2: undefined symbol: ngx_http_apisix_set_gzip
stack traceback:
        [C]: in function '__index'
        /usr/local/apisix/resty/apisix/response.lua:28: in function 'set_gzip'
        /usr/local/apisix/apisix/plugins/gzip.lua:154: in function 'phase_func'
        /usr/local/apisix/apisix/plugin.lua:708: in function 'run_plugin'
        /usr/local/apisix/apisix/plugin.lua:741: in function 'run_global_rules'
        /usr/local/apisix/apisix/init.lua:335: in function 'common_phase'
        /usr/local/apisix/apisix/init.lua:621: in function 'http_header_filter_phase'
        header_filter_by_lua:2: in main chunk while reading response header from upstream, client: 172.16.1.1, server: _, request: "GET /doc.html HTTP/2.0", upstream: "http://172.16.1.1:80/doc.html", host: "remot1.argesvideocloud.com:9443"









快疯掉了。。。。。。。。。。   好不容找到   https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/  放到 了  /ust/local/apisix/resty/ 目录下。 然后就开始报错。

apisix.lua version  2.10.2

**Reply by xlgo** [reply]
nginx version: openresty/1.19.9.1

**Reply by juzhiyuan** [reply]
Hi @xlgo, don't worry, please provide information like:

1. Which Apache APISIX version are you using?
2. After deploying, what changes you have made to it?
3. How to reproduce your issue? (steps)

**Reply by xlgo** [reply]
>  
apisix.lua version 2.10.2

插件里打开了  gzip plugin，默认配置，一直没有生效。
发现有错误，提示需要  core.log.error("need to build APISIX-OpenResty to support setting gzip")  
然后就找到了  https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/  放到目录下
restart 之后  就报错了。禁用插件后正常。

https://apisix.apache.org/docs/apisix/how-to-build
我是按照 上面地址安装的 CentOS 7

重视我的问题啊，主要是看性能不错，配置方便。 就想配置个 gzip 而已。
我没怎么搞过服务器这块，主要是 java 开发。所以也改不了啥东西。   
报错信息还是  看了lua入门之后才找到 需要什么的。

***我仅仅是想用它，所有插件的功能。求大哥给☞个路***
-------------------- 来自百度翻译  for fanyi.baidu.com------------
The gzip plugin is opened in the plug-in. The default configuration has not taken effect.

If an error is found, the prompt needs core log. error("need to build APISIX-OpenResty to support setting gzip")

Then I found it https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/ Put it in the directory

An error is reported after restart. Normal after disabling the plug-in.



https://apisix.apache.org/docs/apisix/how-to-build

I installed CentOS 7 according to the above address



Pay attention to my problem. It mainly depends on good performance and convenient configuration. Just want to configure a gzip.

I haven't done much with servers, mainly java development. So I can't change anything.

After reading Lua's introduction, you can find what you need.
***I just want to use it, the functions of all plug-ins，求 big brother give me ☞ a way***

**Reply by spacewander** [reply]
> > 
> 
> apisix.lua version 2.10.2
> 
> 插件里打开了 gzip plugin，默认配置，一直没有生效。 发现有错误，提示需要 core.log.error("need to build APISIX-OpenResty to support setting gzip") 然后就找到了 [api7/apisix-nginx-module@`main`/lib/resty/](https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/) 放到目录下 restart 之后 就报错了。禁用插件后正常。
> 
> [apisix.apache.org/docs/apisix/how-to-build](https://apisix.apache.org/docs/apisix/how-to-build) 我是按照 上面地址安装的 CentOS 7
> 
> 重视我的问题啊，主要是看性能不错，配置方便。 就想配置个 gzip 而已。 我没怎么搞过服务器这块，主要是 java 开发。所以也改不了啥东西。 报错信息还是 看了lua入门之后才找到 需要什么的。
> 
> _**我仅仅是想用它，所有插件的功能。求大哥给☞个路**_ -------------------- 来自百度翻译 for fanyi.baidu.com------------ The gzip plugin is opened in the plug-in. The default configuration has not taken effect.
> 
> If an error is found, the prompt needs core log. error("need to build APISIX-OpenResty to support setting gzip")
> 
> Then I found it [api7/apisix-nginx-module@`main`/lib/resty/](https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/) Put it in the directory
> 
> An error is reported after restart. Normal after disabling the plug-in.
> 
> [apisix.apache.org/docs/apisix/how-to-build](https://apisix.apache.org/docs/apisix/how-to-build)
> 
> I installed CentOS 7 according to the above address
> 
> Pay attention to my problem. It mainly depends on good performance and convenient configuration. Just want to configure a gzip.
> 
> I haven't done much with servers, mainly java development. So I can't change anything.
> 
> After reading Lua's introduction, you can find what you need. _**I just want to use it, the functions of all plug-ins，求 big brother give me ☞ a way**_

The APISIX OpenResty is a mix of C/Lua, so copying Lua is not enough.
Since you are using CentOS, you can install APISIX by rpm or docker.
https://apisix.apache.org/zh/docs/apisix/how-to-build#%E9%80%9A%E8%BF%87-rpm-%E4%BB%93%E5%BA%93%E5%AE%89%E8%A3%85%EF%BC%88centos-7%EF%BC%89
https://apisix.apache.org/zh/docs/apisix/how-to-build#%E9%80%9A%E8%BF%87-docker-%E5%AE%89%E8%A3%85

They will set up the APISIX OpenResty for you.

**Reply by ccwall3** [reply]
> > > 
> > 
> > 
> > apisix.lua version 2.10.2
> > 插件里打开了 gzip plugin，默认配置，一直没有生效。 发现有错误，提示需要 core.log.error("need to build APISIX-OpenResty to support setting gzip") 然后就找到了 [api7/apisix-nginx-module@`main`/lib/resty/](https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/) 放到目录下 restart 之后 就报错了。禁用插件后正常。
> > [apisix.apache.org/docs/apisix/how-to-build](https://apisix.apache.org/docs/apisix/how-to-build) 我是按照 上面地址安装的 CentOS 7
> > 重视我的问题啊，主要是看性能不错，配置方便。 就想配置个 gzip 而已。 我没怎么搞过服务器这块，主要是 java 开发。所以也改不了啥东西。 报错信息还是 看了lua入门之后才找到 需要什么的。
> > _**我仅仅是想用它，所有插件的功能。求大哥给☞个路**_ -------------------- 来自百度翻译 for fanyi.baidu.com------------ The gzip plugin is opened in the plug-in. The default configuration has not taken effect.
> > If an error is found, the prompt needs core log. error("need to build APISIX-OpenResty to support setting gzip")
> > Then I found it [api7/apisix-nginx-module@`main`/lib/resty/](https://github.com/api7/apisix-nginx-module/tree/main/lib/resty/) Put it in the directory
> > An error is reported after restart. Normal after disabling the plug-in.
> > [apisix.apache.org/docs/apisix/how-to-build](https://apisix.apache.org/docs/apisix/how-to-build)
> > I installed CentOS 7 according to the above address
> > Pay attention to my problem. It mainly depends on good performance and convenient configuration. Just want to configure a gzip.
> > I haven't done much with servers, mainly java development. So I can't change anything.
> > After reading Lua's introduction, you can find what you need. _**I just want to use it, the functions of all plug-ins，求 big brother give me ☞ a way**_
> 
> The APISIX OpenResty is a mix of C/Lua, so copying Lua is not enough. Since you are using CentOS, you can install APISIX by rpm or docker. https://apisix.apache.org/zh/docs/apisix/how-to-build#%E9%80%9A%E8%BF%87-rpm-%E4%BB%93%E5%BA%93%E5%AE%89%E8%A3%85%EF%BC%88centos-7%EF%BC%89 https://apisix.apache.org/zh/docs/apisix/how-to-build#%E9%80%9A%E8%BF%87-docker-%E5%AE%89%E8%A3%85
> 
rpm install  get  error

> They will set up the APISIX OpenResty for you.
[root@DEV-APIGATEWAY output]# ls
apisix-base-1.19.9.1-0.el7.x86_64.rpm

[root@DEV-APIGATEWAY output]# rpm -ivh  apisix-base-1.19.9.1-0.el7.x86_64.rpm 
error: Failed dependencies:
        openresty conflicts with apisix-base-1.19.9.1-0.el7.x86_64

**Reply by juzhiyuan** [reply]
Hi @ccwall3, that's because apisix-base is based on OpenResty, so it conflicts. Have a try by `yum remove openresty`?

**Reply by ccwall3** [reply]
i have done it (remove the openresty,  install apisix-base）
now it get new error , looks like  missing something


error.log:2022/01/06 19:02:29 [warn] 8872#8872: *4272 [lua] real-ip.lua:121: phase_func(): missing real address, client: 10.8.171.254, server: _, request: "POST /backend-auth/user/group/list HTTP/1.1", host: "devapi.x.com"
error.log:2022/01/06 19:04:05 [warn] 8871#8871: *9995 [lua] real-ip.lua:121: phase_func(): missing real address, client: 10.8.171.254, server: _, request: "GET /backend-auth/login HTTP/1.1", host: "devapi.x.com"
error.log:2022/01/06 19:06:15 [warn] 8871#8871: *15478 [lua] real-ip.lua:121: phase_func(): missing real address, client: 10.8.171.254, server: _, request: "GET /backend-auth/login HTTP/1.1", host: "devapi.x.com"

**Reply by juzhiyuan** [reply]
It looks like a new issue, please create a standalone issue or discussion.
